sql/2021/pwa/sw_events.sql

```diff
@@ -0,0 +1,66 @@
+#standardSQL
+# SW events
+
+CREATE TEMPORARY FUNCTION getSWEvents(swEventListenersInfo ARRAY<STRING>, swPropertiesInfo ARRAY<STRING>)
+RETURNS ARRAY<STRING> LANGUAGE js AS '''
+try {
+  return [...new Set([...swEventListenersInfo ,...swPropertiesInfo])];
+} catch (e) {
+  return [e];
+}
+''';
+
+CREATE TEMPORARY FUNCTION parseField(field STRING)
+RETURNS ARRAY<STRING> LANGUAGE js AS '''
+ try {
+     if(field == '[]' || field == '') {
+         return [];
+     }
+     var parsedField = Object.values(JSON.parse(field));
+     if (typeof parsedField != 'string') {
+         parsedField = parsedField.toString();
+     }
+     parsedField = parsedField.trim().split(',');
+     return parsedField;
+ } catch (e) {
+   return [e];
+ }
+ ''';
+
+SELECT
+  _TABLE_SUFFIX AS client,
+  event,
+  COUNT(DISTINCT url) AS freq,
+  total,
+  COUNT(DISTINCT url) / total AS pct
+FROM
+  `httparchive.pages.2021_07_01_*`,
+  UNNEST(getSWEvents(parseField(JSON_EXTRACT(payload, '$._pwa.swEventListenersInfo')), parseField(JSON_EXTRACT(payload, '$._pwa.swPropertiesInfo')))) AS event
```

Comment:
Any reason why you don't pass the JSON_EXTRACT'd value of `$._pwa`, which should be relatively small, and extract the listeners/properties arrays in the UDF?

Reply:
Not really, this it just reflects the way the query was built up. Rewrote.